### PR TITLE
Propagate signature-modifying decorators to pyplot wrappers.

### PR DESCRIPTION
**Pull request by anntzer**
## PR Summary

alternative to #14130, see discussion there; attn @timhoffm 

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/tests/test_pyplot.py [hunk 1]
```diff
@@ -36,6 +37,27 @@ def test_pyplot_up_to_date():
...
+
+    func(None)
+    func(new=None)
+    func(None, kwo=None)
+    func(new=None, kwo=None)
+    assert not recwarn
+    with pytest.warns(MatplotlibDeprecationWarning):
+        func(old=None)
+    with pytest.warns(MatplotlibDeprecationWarning):
+        func(None, None)
```
[the remark this answers is not in the document]
Comment (anntzer) [reply lib/matplotlib/tests/test_pyplot.py]: indeed; this also revealed a nasty interaction between rename_parameter and make_keyword_only (using signature.bind is overly restrictive when the underlying function actually also accepts other signatures) -- fixed now.

**Comment by ksunden** [comment]
What is the disposition on this PR (asking in relation to #15049)? Looks like the conflict is in generated code, and can be resolved rather simply.

**Comment by timhoffm** [comment]
There's no final decision between this and #14130 yet. In the end, I don't have a very strong opinion anymore on which one to choose. The important thing is that we get one of them in, so that we can properly rename parameters and propagate that to pyplot. Since this is internal functionality, we could even switch out the solution if we realize the other one was better.

In practice, we need a second positive review so that this can get merged. The confict is indeed trivial to solve by regenerating the code.

**Comment by anntzer** [comment]
rebased

**Comment by QuLogic** [comment]
I want to merge, but I'm wondering if there's an example that would trigger this yet?

**Comment by ksunden** [comment]
@QuLogic I think there are others, but one such example is the `s` argument to `plt.annotate` being changed to `text` (See #15049, as that change is also required for this example)

The argument (and docstring) were changed in the actual implementation, but pyplot still has the old kwarg (with the new docstring)

**Pull request merged**